### PR TITLE
Overhaul resource duplication

### LINKED ISSUE #74918: Resource.duplicate(true) doesn't duplicate subresources stored in Array or Dictionary properties
### Godot version

4.0.stable

### System information

Manjaro Linux

### Issue description

This behavior is poorly documented, and it breaks expectations. It is mentioned in the docs only under `Array.duplicate()`, but not `Dictionary` class, or `Resource.duplicate()`.

This affects also "Make Unique" and "Make Unique (Recursive)" for custom resources.

Here's the problematic code:
https://github.com/godotengine/godot/blob/fc7adaab7b3856a7831d402ea2bbb27efe7b7d8a/core/variant/variant_setget.cpp#L1889-L1900

### Steps to reproduce

For simple code see reproduction project.

To reproduce on your own:
1) Have a resource with a property of type Array[SomeOtherResource]
2) One of the following:
2a) "Make Unique" on parent resource in inspector.
2b) Or duplicate(true) from code.
3) Check if SomeOtherResource items inside array are unique.

### Minimal reproduction project

[res_bug_reproduction.zip](https://github.com/godotengine/godot/files/10972464/res_bug_reproduction.zip)

### LINKED ISSUE #96627: Dictionary.duplicate(true) still does not make a recursive deep copy
### Tested versions

Personal experience: 4.3
I have also seen [documentation of this bug from over 4 years ago](https://github.com/godotengine/godot/issues/37162), as well as some [more recent reports](https://github.com/mphe/SmartShape2D/issues/32).

### System information

Windows 11

### Issue description

The documentation for the Dictionary class states "If deep is true, inner Dictionary and [Array](https://docs.godotengine.org/en/stable/classes/class_array.html#class-array) keys and values are also copied, **recursively**." for the duplicate method. However, the objects in dictionaries created via the Dictionary.duplicate(true) function still act as references to those in the duplicated dictionary, implying either that a shallow copy or a non-recursive deep copy was made, instead of the recursive deep copy proscribed by the documentation. 

### Steps to reproduce

Create a dictionary. Put an array of objects passed by reference into the dictionary. Call duplicate(true) on the dictionary. Alter an object in the array in the result of the duplicate function. Print or otherwise display the object it corresponds to in the original dictionary. It will appear as if a shallow copy had been made, despite described behavior specifying a deep copy.

### Minimal reproduction project (MRP)

I do not currently have access to a Godot editor to make a MRP. I hope the issue is simple enough fort my description to be sufficient.

### PR DESCRIPTION
**Pull request by RandomShaper**
Fixes #74918.
**2025-05-19:** Fixes #96627.

## TL;DR
This unifies the different paths in Godot for resource duplication so they all have the same base behavior and only differ in what is predictable. Also, an extremely comprehensive set of tests is added.

## Long description

Each commit takes a self-contained step towards the overall goal. Quoting their messages for the details:

### Overhaul `Resource::duplicate_for_local_scene()`

- Serves as a first step for future refactors.
- Code is simpler.
- Algorithm is more efficient: instead of two passes (dumb copy + resolve copies), it's single-pass.
- Now obeys `PROPERTY_USAGE_NEVER_DUPLICATE`.
- Now handles deep self-references (the resource to be duplicated being referenced somewhere deep).

### Overhaul `Resource::duplicate()` (**Edited 2025-01-21.**)

Thanks to a refactor, `Resource::duplicate_for_local_scene()` and `Resource::duplicate()` are now both users of the same, parametrized, implementation.

`Resource::duplicate()` now honors deepness in a more consistent and predictable fashion. `Resource::duplicate_deep()` is added (instead of just adding a parameter to the former, for compatibility needs).

The behavior after this change is as follows:
  - Deep (`deep=true`, formerly `subresources=true`):
    - Previously, only resources found as direct property values of the one to copy would be, recursively, duplicated.
    - Now, in addition, arrays and dictionaries are walked so the copy is truly deep, and only local subresources found across are copied.
    - Previously, subresources would be duplicated as many times as being referenced throughout the main resource.
    - Now, each subresource is only duplicated once and from that point, a referenced to that single copy is used. That's the enhanced behavior that `duplicate_for_local_scene()` already featured.
    - The behavior with respect to packed arrays is still duplication.
    - Formerly, arrays and dictionaries were recursive duplicated, with resources ignored.
    - Now, arrays and dictionaries are recursive duplicated, with resources duplicated.
    - When doing it through `duplicate_deep()`, there's a` deep_subresources_mode` parameter, with various possibilites to control if no resources are duplicated (so arrays, etc. are, but keeping referencing the originals), if only the internal ones are (resources with no non-local path, the default), or if all of them are. The default is to copy every subresource, just like `duplicate(true)`.
  - Not deep (`deep=false`, formerly `subresources=false`): <a name="resource-shallow"></a>
    - Previously, the first level of resources found as direct property values would be duplicated unconditionally. Packed arrays, arrays and dictionaries were non-recursively duplicated.
    - Now, no subresource found at any level in any form will be duplicated, but the original reference kept instead. Packed arrays, arrays and dictionaries are referenced, not duplicated at all.
    - Now, resources found as values of always-duplicate properties are duplicated, recursively or not matching what was requested for the root call.

This commit also changes what's the virtual method to override to customize the duplication (now it's the protected `_duplicate()` instead of the public `duplicate()`).

### Overhaul `Variant::duplicate()` for resources (**Edited 2025-01-21.**)

This in the scope of a duplication triggered via any type in the `Variant` realm. that is, the following: `Variant` itself, `Array` and `Dictionary`. That includes invoking `duplicate()` from scripts.

A `duplicate_deep(deep_subresources_mode)` method is added to `Variant`, `Array` and `Dictionary` (for compatibility reasons, simply adding an extra parameter was not possible). The default value for it is `RESOURCE_DEEP_DUPLICATE_NONE`, which is like calling `duplicate(true)`.

Remarks:
- The results of copying resources via those `Variant` types are exactly the same as if the copy were initiated from the `Resource` type at C++.
- In order to keep some separation between `Variant` and the higher-level animal which is `Resource`, `Variant` still contains the original code for that, so it's self-sufficient unless there's a `Resource` involved. Once the deep copy finds a `Resource` that has to be copied according to the duplication parameters, the algorithm invokes the `Resource` duplication machinery. When the stack is unwind back to a nesting level `Variant` can handle, `Variant` duplication logic keeps functioning.

While that is good from a responsibility separation standpoint, that would have a caveat: `Variant` would not be aware of the mapping between original and duplicate subresources and so wouldn't be able to keep preventing multiple duplicates.

To avoid that, this commit also introduces a wormwhole, a sharing mechanism by which `Variant` and `Resource` can collaborate in managing the lifetime of the original-to-duplicates map. The user-visible benefit is that the overduplicate prevention works as broadly as the whole `Variant` entity being copied, including all nesting levels, regardless how disconnected the data members containing resources may be across al the nesting levels. In other words, despite the aforementioned division of duties between `Variant` and `Resource` duplication logic, the duplicates map is shared among them. It's created when first finding a `Resource` and, however how deep the copy was working at that point, the map kept alive unitl the stack is unwind to the root user call, until the first step of the recursion.

Thanks to that common map of duplicates, this commit is able to fix the issue that `Resource::duplicate_for_local_scene()` used to ignore overridden duplicate logic.

## Caveats (**Edited 2025-01-21.**)

It may be the case that user projects are relying on the former behaviors. On the one hand, most of the time that means they were taking unreasonable or buggy behaviors as good, so I'd say it's not a big problem to take the all workings away.

## Notes on other duplication entry points (**Added 2025-01-03.**)

In order for the scope of this PR to be better understood and also have a list of potential changes for subsequent PRs, let's be mindful of the following:
- Node duplication (editor and runtime):
  - Not directly modified by this PR. (*)
  - Only duplicates top-level properties.
  - Only duplicates resources in always-duplicate properties.
  - Ignores never-duplicate.
  - Only relevant to child nodes, it only recurses into `Array` properties that are typed as object, ignores `Dictionary`.
- Resource make unique (editor):
  - Single
    - Not directly modified by this PR. (*)
  - Recursive:
    - Not directly modified by this PR. (*)
    - Recurses into nested resources found as values of properties that are _not_ never-duplicate. Ignores `Dictionary` and `Array`.
    - Ignores always-duplicate.

*: All of the workflows above use shallow duplication. Therefore, even if not directly modified by this PR, each individual resource duplication is affected by the [behavior changes to shallow resource duplication](#resource-shallow). (**Added 2025-05-21**): In order to keep the current behavior of resouces assigned to properties not being duplicated at all, a check is added to the `Node` duplication algorithm.

---

I've tested this locally and also the test suite is pretty comprehensive. Nonetheless, some extra testing is more than welcome.

### CODE REVIEW

**Comment by RedMser** [comment]
I tested both `resource_local_to_scene` and `resource.duplicate(true)` with sub-resources, arrays and dictionaries, using [this MRP](https://github.com/godotengine/godot/issues/71243). I didn't notice any unexpected regressions compared to before the PR.

File: doc/classes/Resource.xml [hunk 1]
```diff
@@ -53,12 +53,13 @@
...
 				[param subresources] is usually respected, with the following exceptions:
-				- Subresource properties with the [constant PROPERTY_USAGE_ALWAYS_DUPLICATE] flag are always duplicated.
+				- Subresource properties with the [constant PROPERTY_USAGE_ALWAYS_DUPLICATE] flag are always duplicated (recursively).
 				- Subresource properties with the [constant PROPERTY_USAGE_NEVER_DUPLICATE] flag are never duplicated.
-				- Subresources inside [Array] and [Dictionary] properties are never duplicated.
 				[b]Note:[/b] For custom resources, this method will fail if [method Object._init] has been defined with required parameters.
+				[b]Note:[/b] When duplicaating with [param subresources] set to [code]true[/code], each resource found, including the one on which this method is called, will be only duplicated once and referenced as many times as needed in the duplicate. For instance, if you are duplicating resource A that happens to have resource B referenced twice, you'll get a new resource A' referencing a new resource B' twice.
```
Comment (Mickeon) [inline doc/classes/Resource.xml:62]: Other than the typo fix, I have nothing right now. But, the example given may make it sound more confusing than it actually is in practice.
```suggestion
				[b]Note:[/b] When duplicating with [param subresources] set to [code]true[/code], each resource found, including the one on which this method is called, will be only duplicated once and referenced as many times as needed in the duplicate. For instance, if you are duplicating resource A that happens to have resource B referenced twice, you'll get a new resource A' referencing a new resource B' twice.
```

File: doc/classes/Resource.xml [hunk 2]
```diff
@@ -53,12 +53,13 @@
 			<param index="0" name="subresources" type="bool" default="false" />
 			<description>
 				Duplicates this resource, returning a new resource with its [code]export[/code]ed or [constant PROPERTY_USAGE_STORAGE] properties copied from the original.
-				If [param subresources] is [code]false[/code], a shallow copy is returned; nested resources within subresources are not duplicated and are shared with the original resource (with one exception; see below). If [param subresources] is [code]true[/code], a deep copy is returned; nested subresources will be duplicated and are not shared (with two exceptions; see below).
+				If [param subresources] is [code]false[/code], a shallow copy is returned; nested resources within subresources are not duplicated and are shared with the original resource (with one exception; see below).
+				If [param subresources] is [code]true[/code], a deep copy is returned; nested subresources will be duplicated and are not shared (with two exceptions; see below). Nested arrays and dictionaries will also be recursively duplicated. See [method Array.duplicate] and [method Dictionary.duplicate] for more information.
```
[the remark this answers is not in the document]
Comment (RandomShaper) [reply doc/classes/Resource.xml:57]: I'm choosing the second option.

File: doc/classes/Resource.xml (hunk 1 above)
Comment (RandomShaper) [reply doc/classes/Resource.xml]: Suggestion applied, by the way. Marking as resolved for now, so I can track better what I have already addressed. Feel free to unresolve if you come up with something else related to this.

**Comment by RandomShaper** [comment]
Dear reviewers. I'm sorry, but I have needed to update this PR with a bunch of changes to support additional duplicate modes, for greater flexibility and compatibility preservation. I've updated all the commit messages as well as the description of the PR, which includes them.

**Comment by AdriaandeJongh** [comment]
I so very appreciate this PR and really look forward to these changes making it into the engine! Thank you for addressing this. In the meantime I wanted to mention that I am _very_ lost in the differences between `duplicate(deep=false)`, `duplicate(deep=true)`, and `duplicate_deep()`. You mention that `duplicate_deep()` is added for compatibility needs, but if `duplicate()` already takes a `deep` parameter, what is the specific compatibility need here? My intuition is that a single `duplicate()` function with all the necessary parameters is the least confusing for all users.

**Comment by HampusHauffman** [comment]
What is needed to get this merged?

**Comment by Mickeon** [comment]
A look from **Core** maintainers. It's a pretty extensive change so it may need a lot of scrutiny.

**Comment by RandomShaper** [comment]
> In the meantime I wanted to mention that I am _very_ lost in the differences between `duplicate(deep=false)`, `duplicate(deep=true)`, and `duplicate_deep()`. You mention that `duplicate_deep()` is added for compatibility needs, but if `duplicate()` already takes a `deep` parameter, what is the specific compatibility need here? My intuition is that a single `duplicate()` function with all the necessary parameters is the least confusing for all users.

The problem is `Variant`. It's not possible to extend the signature of `duplicate()` without breaking compatiblity with GDExtension. At the moment, `Variant` doesn't feature a system to provide compatibility methods, and such a thing would most likely be detrimental to performance across all `Variant` calls anyway. Therefore, had no option but adding a brand new method and, for the sake of clarity, replicating the approach in `Resource`, where it would have been possible to hae a compatibility method.

**Comment by KoBeWi** [comment]
I tested the MRP from the issue and the array is empty after duplication.

**Comment by RandomShaper** [comment]
@KoBeWi, thanks a lot for testing this. There was indeed a bug involving scripted properties. Given the new code deals directly with resources, it needed special treatment for those properties so the assignments are not rejected; namely:
- `script` property has to be copied first, so the targe properties exist.
- Duplication of typed arrays and dictionaries has to transfer the type to the copies.

I have fixed that in my latest push. A subsequent push will be made, rebased and with formatting fixed.

File: core/variant/variant_call.cpp [hunk 3]
```diff
@@ -2334,6 +2334,7 @@ static void _register_variant_builtin_methods_misc() {
 	bind_method(Dictionary, keys, sarray(), varray());
 	bind_method(Dictionary, values, sarray(), varray());
 	bind_method(Dictionary, duplicate, sarray("deep"), varray(false));
+	bind_method(Dictionary, duplicate_deep, sarray("deep_subresources_mode"), varray(false, RESOURCE_DEEP_DUPLICATE_INTERNAL));
```
Comment (KoBeWi) [inline core/variant/variant_call.cpp:2337]: Why is there `false` in the varray?

File: core/variant/variant_call.cpp [hunk 4]
```diff
@@ -2392,6 +2393,7 @@ static void _register_variant_builtin_methods_array() {
 	bind_method(Array, bsearch_custom, sarray("value", "func", "before"), varray(true));
 	bind_method(Array, reverse, sarray(), varray());
 	bind_method(Array, duplicate, sarray("deep"), varray(false));
+	bind_method(Array, duplicate_deep, sarray("deep_subresources_mode"), varray(RESOURCE_DEEP_DUPLICATE_NONE));
```
Comment (KoBeWi) [inline core/variant/variant_call.cpp:2396]: Shouldn't this also use RESOURCE_DEEP_DUPLICATE_INTERNAL?

**Comment by KoBeWi** [comment]
> I have fixed that in my latest push.

Did you add a test for it? I see this PR already adds some tests, but they didn't cover that case.

File: core/variant/variant_call.cpp (hunk 4 above)
Comment (RandomShaper) [reply core/variant/variant_call.cpp:2396]: Good catch. Fixed in the upcoming push.

**Comment by RandomShaper** [comment]
> Did you add a test for it? I see this PR already adds some tests, but they didn't cover that case.

Included in the upcoming push.

**Comment by RandomShaper** [comment]
Rebased to adapt to the new `GDSOFTCLASS`.

**Comment by KoBeWi** [comment]
One of the tests is crashing.

**Comment by RandomShaper** [comment]
The tests were using `GDREGISTER_INTERNAL_CLASS` to instantiate the custom test resources, but that's no longer allowed. I've pushed with a change of that to `GDREGISTER_CLASS`.

**Comment by RandomShaper** [comment]
Aargh, forgot to re-run doctool. Hopefully all good now.

**Review by KoBeWi — APPROVED** [review]

**Comment by RandomShaper** [comment]
@dsnopek, yes, they are subclasses of `Resource`.
 I used `GDREGISTER_INTERNAL_CLASS` with the idea that it would be the lowest degree of exposure the tests could work with. So, there was not really a proper use case, but just an attempt at purism. Nothing is actually lost by using `GDREGISTER_CLASS`.

**Comment by RandomShaper** [comment]
Rebased.

**Comment by RandomShaper** [comment]
Re-rebased.

**Comment by mihe** [comment]
@RandomShaper There seems to be some kind of regression here.

Open this MRP and duplicate the label in the main scene, and you'll find that it duplicates the `ExtResource` assigned to `Label1/theme_override_fonts/font`: [font-duplication.zip](https://github.com/user-attachments/files/20300710/font-duplication.zip)

**Comment by RandomShaper** [comment]
Fixed. I've also added an update to the latest part of the PR description. Quoting it here:
> In order to keep the current behavior of resouces assigned to properties not being duplicated at all, a check is added to the Node duplication algorithm.

File: doc/classes/Array.xml [hunk 5]
```diff
@@ -332,7 +332,16 @@
 			<param index="0" name="deep" type="bool" default="false" />
 			<description>
 				Returns a new copy of the array.
-				By default, a [b]shallow[/b] copy is returned: all nested [Array] and [Dictionary] elements are shared with the original array. Modifying them in one array will also affect them in the other.[br]If [param deep] is [code]true[/code], a [b]deep[/b] copy is returned: all nested arrays and dictionaries are also duplicated (recursively).
+				By default, a [b]shallow[/b] copy is returned: all nested [Array], [Dictionary], and [Resource] elements are shared with the original array. Modifying them in one array will also affect them in the other.
```
[the remark this answers is not in the document]
Comment (RandomShaper) [reply doc/classes/Array.xml:335]: Regarding this and the analogous suggestion for `Dictionary`, to me, this suggestion introduces an assymmetry. I mean, my original sentence talks aboutthe items; you modify an item in the source array and that will make that item result modified also in the copy of the array, because it's the same. The suggestion sounds to me like, you modify an item in the source array and that makes the copy of the array result modified. I'm not sure I'm getting my message across.

In other words, my sentence is about item here and item there, whereas the suggestion is about item here and the 'there'.

[the remark this answers is not in the document]
Comment (AThousandShips) [reply doc/classes/Array.xml:335]: The sentence seems to me to not be complete, I'd suggest:
```suggestion
				By default, a [b]shallow[/b] copy is returned: all nested [Array], [Dictionary], and [Resource] elements are shared with the original array. Modifying the elements in one array will also affect them in the other.
```

**Review by Calinou — APPROVED** [review]
Tested locally (by editing the 3D platformer demo project), it works as expected. Code looks good to me.

I like the addition of `duplicate_deep()` for more explicit code.

**Comment by HampusHauffman** [comment]
Thank you!
